README.md
# Biml Support

Very basic language support for [Business Intelligence Markup Language](https://www.varigence.com/Biml) (`.biml`) files.

## Features

Provides syntax highlighting for the Business Intelligence Markup Language.

![Screenshot of functionality](./assets/screenshot1.png)

## Known Issues

Currently only `C#` embedded highlighting is supported.

## Release Notes

### 1.0.1

Fix an issue where back-to-back blocks caused invalid highlighting.

### 1.0.0

Initial release.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.1] - 2020-05-14

-   Fix an issue where back-to-back blocks caused invalid highlighting.

## [1.0.0] - 2019-07-19

-   Initial release with basic embedded syntax highlighting.


src/extension.ts
import * as vscode from "vscode";

export function activate(context: vscode.ExtensionContext) {
	const bracketDecorationType = vscode.window.createTextEditorDecorationType({
		light: {
			backgroundColor: "rgba(255, 100, 0, .2)"
		},
		dark: {
			backgroundColor: "rgba(0, 100, 255, .2)"
		}
	});

	const codeBlockDecorationType = vscode.window.createTextEditorDecorationType(
		{
			light: {
				backgroundColor: "rgba(100,100,100,0.1)"
			},
			dark: {
				backgroundColor: "rgba(220,220,220,0.1)"
			}
		}
	);

	let activeEditor = vscode.window.activeTextEditor;
	if (activeEditor) {
		triggerUpdateDecorations();
	}

	vscode.window.onDidChangeActiveTextEditor(
		editor => {
			activeEditor = editor;
			if (editor) {
				triggerUpdateDecorations();
			}
		},
		null,
		context.subscriptions
	);

	vscode.workspace.onDidChangeTextDocument(
		event => {
			if (activeEditor && event.document === activeEditor.document) {
				triggerUpdateDecorations();
			}
		},
		null,
		context.subscriptions
	);

	var timeout = null;
	function triggerUpdateDecorations() {
		if (timeout) {
			clearTimeout(timeout);
		}
		timeout = setTimeout(updateDecorations, 200);
	}

	function updateDecorations() {
		if (!activeEditor || !isCandidateFile(activeEditor)) {
			return;
		}
		const regEx = /(<#@|<#\+|<#=|<#)|(#>)+/g;
		const text = activeEditor.document.getText();
		const brackets: vscode.DecorationOptions[] = [];
		let match;
		while ((match = regEx.exec(text))) {
			const startPos = activeEditor.document.positionAt(match.index);
			const endPos = activeEditor.document.positionAt(
				match.index + match[0].length
			);
			const decoration = {
				range: new vscode.Range(startPos, endPos),
				hoverMessage: ""
			};

			brackets.push(decoration);
		}

		const blocks: vscode.DecorationOptions[] = [];

		let index = 0;
		let max = brackets.length;
		brackets.forEach(element => {
			if (index + 1 < max) {
				const start = brackets[index];
				const end = brackets[index + 1];

				const decoration: vscode.DecorationOptions = {
					range: new vscode.Range(start.range.end, end.range.start),
					hoverMessage: ""
				};

				blocks.push(decoration);
			}

			index += 2;
		});

		activeEditor.setDecorations(bracketDecorationType, brackets);
		activeEditor.setDecorations(codeBlockDecorationType, blocks);
	}

	function isCandidateFile(editor: vscode.TextEditor): boolean {
		return editor.document.languageId === "biml";
	}
}


out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vscode = require("vscode");
function activate(context) {
    var bracketDecorationType = vscode.window.createTextEditorDecorationType({
        light: {
            backgroundColor: "rgba(255, 100, 0, .2)"
        },
        dark: {
            backgroundColor: "rgba(0, 100, 255, .2)"
        }
    });
    var codeBlockDecorationType = vscode.window.createTextEditorDecorationType({
        light: {
            backgroundColor: "rgba(100,100,100,0.1)"
        },
        dark: {
            backgroundColor: "rgba(220,220,220,0.1)"
        }
    });
    var activeEditor = vscode.window.activeTextEditor;
    if (activeEditor) {
        triggerUpdateDecorations();
    }
    vscode.window.onDidChangeActiveTextEditor(function (editor) {
        activeEditor = editor;
        if (editor) {
            triggerUpdateDecorations();
        }
    }, null, context.subscriptions);
    vscode.workspace.onDidChangeTextDocument(function (event) {
        if (activeEditor && event.document === activeEditor.document) {
            triggerUpdateDecorations();
        }
    }, null, context.subscriptions);
    var timeout = null;
    function triggerUpdateDecorations() {
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(updateDecorations, 200);
    }
    function updateDecorations() {
        if (!activeEditor || !isCandidateFile(activeEditor)) {
            return;
        }
        var regEx = /(<#@|<#\+|<#=|<#)|(#>)+/g;
        var text = activeEditor.document.getText();
        var brackets = [];
        var match;
        while ((match = regEx.exec(text))) {
            var startPos = activeEditor.document.positionAt(match.index);
            var endPos = activeEditor.document.positionAt(match.index + match[0].length);
            var decoration = {
                range: new vscode.Range(startPos, endPos),
                hoverMessage: ""
            };
            brackets.push(decoration);
        }
        var blocks = [];
        var index = 0;
        var max = brackets.length;
        brackets.forEach(function (element) {
            if (index + 1 < max) {
                var start = brackets[index];
                var end = brackets[index + 1];
                var decoration = {
                    range: new vscode.Range(start.range.end, end.range.start),
                    hoverMessage: ""
                };
                blocks.push(decoration);
            }
            index += 2;
        });
        activeEditor.setDecorations(bracketDecorationType, brackets);
        activeEditor.setDecorations(codeBlockDecorationType, blocks);
    }
    function isCandidateFile(editor) {
        return editor.document.languageId === "biml";
    }
}
exports.activate = activate;
//# sourceMappingURL=extension.js.map